### ISSUE TITLE
DualPrompt: Complementary Prompting for Rehearsal-free Continual Learning

### ISSUE THREAD

**Issue by HYUNMIN-HWANG** [One-Day-One-Paper/One-Day-One-Paper#294]
Continual learning에서 catastrophic forgetting을 해결하기 위해서 그동안 많은 논문에서 rehearsal buffer를 만들었다, past task에 대한 학습을 저장하기 위한 buffer를 만든 것인데 1) sensitive to the size of the buffer 2) privacy 정보 고려 못함 3) memory budget is highly constrained 하다는 문제가 있다.

따라서 해당 문제에서는 rehearsal buffer 없이 prompt를 사용해  Continual learning을 할 수 있는 모델을 제안했다.

- we present DualPrompt, a rehearsal-free continual learning approach to explicitly learn two sets of disjoint prompt spaces, G(eneral)-Prompt and E(xpert)-Prompt, that encode task-invariant and task-specific instructions, respectively.
- we introduce Split ImageNet-R, a new CL benchmark based on ImageNet-R [16] to the community.

## DualPrompt
![Untitled](https://user-images.githubusercontent.com/70581043/192836961-d7f735cb-b138-496c-b98b-97abd3615864.png)

### 1. G-Prompt & E-Prompt

- General prompt (task-invariant)
    
    $h_g^{(i)}=f_{prompt}(g,h^{(i)})$
    
    $g\in \mathbb{R}^{L_g\times D}$ : $L_g$ sequence length, D embedding dimension
    
    G-prompt와 $i$-th MSA layer를 붙인다.
    
    - $f_{prompt}$ : how to attach the prompts to the hidden embeddings
- Expert prompt (task-specific)
    
    $h_e^{(j)}=f_{prompt}(e_t,h^{(j)})$
    
    $E=\{e_t\}_{t=1}^T$ : set of task-dependent parameters
    
    $e\in \mathbb{R}^{L_e\times D}$ : $L_e$ sequence length, D embedding dimension
    
    $T$ : total number of tasks
    
    each $e_t$ is associated with a task-specific key $k_t\in\mathbb{R}^D$ → learnable parameter, capture representative features of a task
    
    E-prompt와 $j$-th MSA layer를 붙인다.
    
- matching loss
    
    $k_t$ becomes “closer” to examples from the $t$-th task than other keys.
    
    task keys에 match 되는 query function $q$ → 여기에 대응되는 E-prompt를 선택
    
    $\mathcal{L}_{match}(x,k_t) = \gamma(q(x),k_t), x\in D_t$
    
    - $\gamma$  : cosine similarity
    - $q(x)=f(x)[0]$ ([class] token)
    
    argmin_t가 되는 방향을 찾는다.
    

## 2. prompt attaching

- where : Decoupled prompt positions
    - G-prompt $g=\{g^{(l)}\}_{l=start_g}^{end_g}$ 이거랑 attach start_g-th ~ end_g-th  MSA layer
    - E-prrompt $e_t=\{e_t^{(l)}\}_{l=start_e}^{end_e}$ 이거랑 attach start_e-th ~ end_e-th MSA layer
    - 이 둘은 절대 overlapping 안한다.
- How : Configurable prompting function
    - Prompt Tuning (Pro-T) : prompt paramter p를 h의 q,k,v 모두에 붙인다.
        
         $f_{prompt}^{Pro-T}(p,h)=MSA([p;h_Q],[p;h_K],[p;h_V])$
        
    - Prefix Tuning (Pre-T) : prompt parameter p를 k,v로 분리하고 각각 해당하는 h에 붙인다.
        
         $f_{prompt}^{Pre-T}(p,h)=MSA(h_Q,[p_k;h_K],[p_v;h_V])$
        
    

## 3. Overal objective

input x → architecture with prompts $f_{g,e_t}$ →classfication $f_\phi$ → cross entropy loss 

$\min_{g,e,k_t,\phi}\mathcal{L}(f_\phi(f_{g,e_t}(x)),y)+\lambda \mathcal{L}_{match}(x,k_t), x\in D_t$

---

### Split ImageNet-R

![Untitled 1](https://user-images.githubusercontent.com/70581043/192837060-a7d002f3-3d4e-4863-a200-376feb69a390.png)

The Split ImageNet-R benchmark is build upon ImageNet-R  by dividing the 200 classes randomly into 10 tasks with 20 classes per task.

장점

- contains classes with different styles
- The significant intra-class diversity
- Pre-trained vision models are useful in practice for many fields

---

## Results
![Untitled 2](https://user-images.githubusercontent.com/70581043/192837088-3c676409-f56e-4dcf-a16a-aed5ad8e358c.png)
